### PR TITLE
Support input/export names in ONNX

### PR DESCRIPTION
**Pull request by zdevito**
This makes it easier to use ONNX graphs in Caffe2 since the input and output nodes will not require finding "magic" constants.

### CODE REVIEW

File: torch/csrc/onnx/export.cpp [hunk 1]
```diff
@@ -256,18 +256,46 @@ static void addAttribute(onnx::NodeProto * n_p, jit::Node * n, jit::Symbol name)
...
+  auto initNames = [&](const node_list & nodes, const std::vector<std::string> & names) {
+    JIT_ASSERT(names.size() <= nodes.size());
+    size_t idx = nodes.size() - names.size();
+    for (auto & name : names) {
+      auto node = nodes.at(idx++);
+      auto it = node_names.find(node);
+      // rarely, output nodes may be repeated
+      // if this happens, we just take the first name we find for it as its name
+      if(it == node_names.end()) {
+        node_names[node] = name;
```
Comment (ezyang) [inline torch/csrc/onnx/export.cpp]: If this case triggers, the export process will silently fail to give a correct name to the output. When a user attempts to run the model from ONNX, they will simply be told that a name (they rightly expect to exist) doesn't exist. You won't be able to debug this situation without looking at the protobuf.

It would be much better to just raise an error in this case.

File: torch/onnx.py [hunk 2]
```diff
@@ -32,12 +39,23 @@ def export(model, args, f, export_params=True, kwargs=None, verbose=False):
         export_params (bool, default True): if specified, all parameters will
             be exported.  Set this to False if you are exporting an
             untrained model.
+        input_names (list, default ()): if specified, a list of strings which
+        will be used in ONNX as the names of the inputs of the model.
+        output_names (list, default ()): if specified, a list of strings which
+        will be used in ONNX as the name of the outputs of the output.
```
Comment (ezyang) [inline torch/onnx.py]: Docs here should mention something about flattening semantics.

File: torch/csrc/jit/python_tracer.cpp [hunk 3]
```diff
@@ -82,13 +82,14 @@ void initPythonTracerBindings(PyObject* module_) {
       ss << *s.graph;
       return ss.str();
     })
-    .def("export", [](TracingState& s, bool verbose) {
+    .def("export", [](TracingState& s,
```
Comment (ezyang) [inline torch/csrc/jit/python_tracer.cpp]: Can we put keyword arguments on these? See http://pybind11.readthedocs.io/en/master/basics.html#keyword-args for how to do it.

Comment (zdevito) [reply torch/csrc/jit/python_tracer.cpp]: that stuff looks like black magic. I tried it before, it didn't work, and I gave up. I think there are part of pybind11 we should avoid using.

File: torch/csrc/onnx/export.cpp (hunk 1 above)
Comment (zdevito) [reply torch/csrc/onnx/export.cpp]: This will make it impossible to use output_names on any network that does this. Is that OK?

File: torch/csrc/onnx/export.cpp [hunk 4]
```diff
@@ -256,18 +256,46 @@ static void addAttribute(onnx::NodeProto * n_p, jit::Node * n, jit::Symbol name)
...
+      }
+    }
+  };
+  initNames(g->inputs(), input_names);
+  initNames(g->outputs(), output_names);
+  auto node_name = [&](Node * n) {
+    auto it = node_names.find(n);
+    if(it != node_names.end())
+      return it->second;
+    return "$" + n->uniqueName();
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/onnx/export.cpp]: Can you suggest an alternative?

File: torch/csrc/onnx/export.cpp (hunk 1 above)
Comment (ezyang) [reply torch/csrc/onnx/export.cpp]: Well, as I mentioned during our check-in today, if we were going to fix this "properly", we'd open a PR to ONNX to add named outputs as an extra piece of metadata in the output. Then we could rightfully only be in the business of assigning to 'env' at binding sites (input is a binding site; output is NOT) and both the "intermediate result returned twice" and the "input returned directly as output" cases would work correctly. But this involves doing the onnx runaround...

Absent that, I think erroring in this case is pretty darn reasonable. I mean, it's not like output_names is going to "work" anyway, if the user actually wants to use the named outputs on the other end.

File: torch/csrc/onnx/export.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/onnx/export.cpp]: Well, there are a few possibilities.

1. Reserve identifiers which start with `[0-9]`. Now you don't need to rename the numeric identifiers with a leading dollar. This is what LLVM does, btw: http://llvm.org/docs/LangRef.html#identifiers
2. Fix up the ONNX printer to not prepend the output with `%`. Perhaps if we want named identifiers to still display with a prefix, we can test if the first character is a `$` or not, and prepend `%` if it is.

[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/onnx/export.cpp]: I think I will just enforce that input_names/output_names must be a valid C identifier. That makes sure it is not equal to a number-based thing, and we can make our uniqueName insert a non-identifier character to prevent overlap with debugNames as well.

File: test/expect/TestONNX.test_batchnorm.expect [hunk 5]
```diff
@@ -7,8 +7,8 @@ node {
   output: "7"
   output: "8"
   output: "9"
-  output: "batch_norm_dead_output_10"
-  output: "batch_norm_dead_output_11"
+  output: "10_batch_norm_dead_output"
```
Comment (ezyang) [inline test/expect/TestONNX.test_batchnorm.expect]: Aw yiss, that reserved namespace ;)

**Comment by ezyang** [comment]
So, this doesn't necessarily block this PR, but I just realized that there is another problem with input/output names in onnx_caffe2, and one that is actually reasonably likely to occur in practice. Suppose that the last operation before output is an inplace operation. When the Caffe2 reference implementation transforms the ONNX proto into Caffe2 proto, it rewrites the output names of inplace ops into the same name as the one that was fed in (since the operation was done inplace.)

Oops! Now you've lost the output name that you carefully put into ONNX.

File: test/test_onnx.py [hunk 6]
```diff
@@ -83,5 +86,17 @@ def test_maxpool(self):
...
+        x = Variable(torch.Tensor([[1, 2], [3, 4]]), requires_grad=True)
+        y = Variable(torch.Tensor([[1, 2], [3, 4]]), requires_grad=True)
+
+        class M(nn.Module):
+            def forward(self, x, y):
+                return -torch.sigmoid(torch.tanh(x * (x + y)))
+
+        f = io.BytesIO()
+        torch.onnx.export(M(), (x, y), f, input_names=["x", "y"], output_names="z")
+        self.assertONNXExpected(f.getvalue())
```
Comment (ezyang) [inline test/test_onnx.py]: It would also be nice to have a test for the error case.

File: torch/onnx.py [hunk 7]
```diff
@@ -13,7 +13,14 @@
 from ._utils import _range
 
 
-def export(model, args, f, export_params=True, kwargs=None, verbose=False):
+def _no_reserved(names):
+    for n in names:
+        if len(n) > 0 and n[0].isdigit():
```
Comment (ezyang) [inline torch/onnx.py]: So, I guess we're not going for the "C-style identifier check" anymore?

File: torch/csrc/autograd/functions/convolution.cpp [hunk 8]
```diff
@@ -55,7 +55,7 @@ auto ConvParams::is_output_padding_neg() const -> bool {
 
 auto ConvParams::is_output_padding_big() const -> bool {
   bool is_big = false;
-  for (int i = 0; i < output_padding.size(); i++) {
+  for (size_t i = 0; i < output_padding.size(); i++) {
```
Comment (apaszke) [inline torch/csrc/autograd/functions/convolution.cpp]: I know it fixes a warning but it will conflict with upstream that has it fixed already

File: torch/csrc/onnx/export.cpp [hunk 9]
```diff
@@ -256,18 +256,48 @@ static void addAttribute(onnx::NodeProto * n_p, jit::Node * n, jit::Symbol name)
...
+  const std::vector<std::string> & input_names,
+  const std::vector<std::string> & output_names,
+  const std::vector<at::Tensor> & initializers) {
+
+  std::unordered_map<Node *, std::string> node_names;
+  auto initNames = [&](const node_list & nodes, const std::vector<std::string> & names) {
+    JIT_ASSERT(names.size() <= nodes.size());
+    size_t idx = nodes.size() - names.size();
+    for (auto & name : names) {
+      auto node = nodes.at(idx++);
```
Comment (apaszke) [inline torch/csrc/onnx/export.cpp]: You have an assert above that guarantees that you won't have out-of-bound accesses

File: torch/csrc/onnx/export.cpp [hunk 10]
```diff
@@ -256,18 +256,48 @@ static void addAttribute(onnx::NodeProto * n_p, jit::Node * n, jit::Symbol name)
...
+
+static void encodeGraph(onnx::GraphProto * p_g, std::shared_ptr<Graph> & g,
+  const std::vector<std::string> & input_names,
+  const std::vector<std::string> & output_names,
+  const std::vector<at::Tensor> & initializers) {
+
+  std::unordered_map<Node *, std::string> node_names;
+  auto initNames = [&](const node_list & nodes, const std::vector<std::string> & names) {
+    JIT_ASSERT(names.size() <= nodes.size());
+    size_t idx = nodes.size() - names.size();
```
Comment (apaszke) [inline torch/csrc/onnx/export.cpp]: Uh our convention for prepending parameters is so weird. I just hit that when writing tests. When people write functional style code inputs usually come before weights, so the traces look completely different 😕

File: torch/csrc/onnx/export.cpp [hunk 11]
```diff
@@ -256,18 +256,48 @@ static void addAttribute(onnx::NodeProto * n_p, jit::Node * n, jit::Symbol name)
...
+      // if this happens, we just fail since using output_names is ambiguous in this case
+      if(it != node_names.end()) {
+        jit::barf("output '%s' and '%s' are the same node, so they cannot have the same output name",
+        name.c_str(), it->second.c_str());
+      }
+      node_names[node] = name;
+    }
+  };
+  initNames(g->inputs(), input_names);
+  initNames(g->outputs(), output_names);
```
Comment (apaszke) [inline torch/csrc/onnx/export.cpp]: Can you move this init to a separate function that returns `node_names`? There's no reason it has to stay in the body of this huge thing

File: torch/csrc/onnx/export.cpp (hunk 10 above)
Comment (ezyang) [reply torch/csrc/onnx/export.cpp]: It's not too late to swap it :)

**Comment by lantiga** [comment]
As suggested by @fmassa on Slack, I'm pointing out a branch:
https://github.com/lantiga/pytorch/tree/blocks
that adds namespace annotations to nodes with just very minor additions to the code.

Example here: https://gist.github.com/lantiga/15ba60f6dbdbc99873f0af94761e9630. There's something weird going on in the graph generated from Test2, but it's likely the way I traverse the IR graph. Also, the way I get input names is superseded by this PR.

Mentioning this here since it would be nice to include annotations to the ONNX export, for debugging/visualization purposes.

File: torch/csrc/onnx/export.cpp (hunk 10 above)
Comment (apaszke) [reply torch/csrc/onnx/export.cpp]: I can't see why

**Comment by apaszke** [comment]
@lantiga we will want to merge this PR anyway. Annotations are gret for inspecting interior of the graph, but it's still useful to allow people to override input/output names because that's the "interface" the model will have when used in other backends

**Comment by apaszke** [comment]
@ezyang can't this be solved by propagating in-place name changes backwards instead of forwards?

**Comment by lantiga** [comment]
@apaszke for sure, my branch is fully orthogonal to this. I mentioned my branch as in a way it provides another piece of the same puzzle.

**Comment by ezyang** [comment]
@apaszke Yes, probably. Also, I also realized that the only inplace op the exporter is currently exporting today is BatchNorm, so this won't blow up until we actually add inplace proper.

**Comment by zdevito** [comment]
I think we should abandon this patch.

Like LLVM, ONNX inputs and outputs are only meaningfully defined based on position. The input and output names are just used inside the function to specify the implementation.  ONNX takes advantage of this, doing renaming for inplace in Caffe2, etc. 

Any attempt to force stable names on ONNX in an ad hoc way is going to just create more problems down the road,  especially if people begin to expect stable naming. 

What can we do instead? I think we should just supplement onnx_caffe2 with a helper functions that let you stably extract the Caffe2 name for a parameter based on its position. Something like `backend.caffe2NameForInput(0)`, `backend.caffe2NameForOutput(0)`. These functions are actually trivial one-liners but we should provide them anyway so people know how to do it.

 We should then flip the input/parameter order so that true inputs come first so people don't have to use `-1` to get the inputs. If we ever add named inputs/outputs proper to ONNX we can then think about adding more functionality here.

**Comment by lantiga** [comment]
@zdevito would debug annotations still make sense to be in ONNX? Even if renaming takes place, the node might retain the annotation. Or a backend might just decide to ignore annotations altogether.

**Comment by zdevito** [comment]
@lantiga: debug annotations are still a good idea and we should continue to pursue that direction by adding annotations to the trace. I just don't want to make guarantees in the API about what the final names will be. For instance, it should be ok if you tell the trace to add a debug name "foo" that it actually names the node "foo.1" because it needed to for some reason. It should also be fine for it to delete the node in an optimization. 

This patch was really only focused on providing input/output names for use in Caffe2, and not the general debugging problem.
